## Project Overview

This project for Trafilea ecommerce has the following objectives 

1. Customer Acquisition & Marketing Efficiency
   
* Customer Acquisition Cost (CAC) Analysis: Identifying the most cost-effective channels for acquiring customers.

* Attribution Modeling: Determining which marketing touchpoints contribute most to conversions.

* Conversion Rate Optimization: Analyzing customer journeys to identify friction points and improve conversion rates.


2. Personalization & Customer Experience Enhan
   cement
* Product Recommendation System: Refining AI-driven recommendations based on browsing and purchase history.

* Customer Segmentation: Clustering customers based on behavior, preferences, and purchasing power for targeted marketing.

* Sentiment Analysis: Analyzing customer reviews and support interactions to identify pain points.



## Data Dictionary 

### leads_qualified
- mql_id: Marketing Qualified Lead unique identifier
- first_contact_date: Date of initial contact with lead
- landing_page_id: ID of landing page where lead originated
- origin: Marketing channel/source of the lead

### leads_closed
- mql_id: Marketing Qualified Lead ID 
- seller_id: ID of seller
- sdr_id: Sales Development Representative ID
- sr_id: Sales Representative ID
- won_date: Date of conversion
- business_segment: Business category
- lead_type: Type of lead
- lead_behaviour_profile: Lead behavior classification
- has_company: Company ownership indicator
- has_gtin: Global Trade Item Number indicator
- average_stock: Average stock level
- business_type: Type of business
- declared_product_catalog_size: Reported product catalog size
- declared_monthly_revenue: Self-reported monthly revenue

### customers
- customer_id: Unique customer identifier
- customer_unique_id: Unique customer identifier
- customer_zip_code_prefix: ZIP code prefix
- customer_city: City location
- customer_state: State location

### sellers
- seller_id: Unique seller identifier
- seller_zip_code_prefix: ZIP code prefix
- seller_city: City location
- seller_state: State location

### orders
- order_id: Unique order identifier
- customer_id: Customer identifier
- order_status: Status of order
- order_purchase_timestamp: Purchase timestamp
- order_approved_at: Approval timestamp
- order_delivered_carrier_date: Carrier delivery date
- order_delivered_customer_date: Customer delivery date
- order_estimated_delivery_date: Estimated delivery date

### order_items
- order_id: Order identifier
- order_item_id: Item identifier within order
- product_id: Product identifier
- seller_id: Seller identifier
- shipping_limit_date: Shipping deadline
- price: Item price
- freight_value: Shipping cost

### order_payments
- order_id: Order identifier
- payment_sequential: Payment sequence number
- payment_type: Type of payment
- payment_installments: Number of installments
- payment_value: Payment amount

### order_reviews
- review_id: Review identifier
- order_id: Order identifier
- review_score: Rating score
- review_comment_title: Review title
- review_comment_message: Review content
- review_creation_date: Review creation date
- review_answer_timestamp: Review response timestamp

### products
- product_id: Product identifier
- product_category_name: Category name in Portuguese
- product_name_lenght: Product name length
- product_description_lenght: Product description length
- product_photos_qty: Number of product photos
- product_weight_g: Weight in grams
- product_length_cm: Length in centimeters
- product_height_cm: Height in centimeters
- product_width_cm: Width in centimeters

### product_category_name_translation
- product_category_name: Category name in Portuguese
- product_category_name_english: Category name in English

### geolocation
- geolocation_zip_code_prefix: ZIP code prefix
- geolocation_lat: Latitude
- geolocation_lng: Longitude
- geolocation_city: City name
- geolocation_state: State name

In [1]:
# import neccessary libraries
import pandas as pd
import numpy as np
import sqlite3
from datetime import datetime

In [2]:
# Connect to my sql database
conn = sqlite3.connect('eCommerce.sqlite')

In [3]:
# Get list of all tables
query = "SELECT name FROM sqlite_master WHERE type='table';"
tables = pd.read_sql_query(query, conn)
tables

name
0   product_category_name_translation
1                             sellers
2                           customers
3                         geolocation
4                         order_items
5                      order_payments
6                       order_reviews
7                              orders
8                            products
9                     leads_qualified
10                       leads_closed

In [4]:
# see all columns in each table
tables = [
    'product_category_name_translation',
    'sellers',
    'customers',
    'geolocation',
    'order_items',
    'order_payments',
    'order_reviews',
    'orders',
    'products',
    'leads_qualified',
    'leads_closed'
]

for table in tables:
    print(f"\n=== {table} ===")
    df = pd.read_sql_query(f"SELECT * FROM {table} LIMIT 1", conn)
    print(df.columns.tolist())



=== product_category_name_translation ===
['product_category_name', 'product_category_name_english']

=== sellers ===
['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state']

=== customers ===
['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']

=== geolocation ===
['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng', 'geolocation_city', 'geolocation_state']

=== order_items ===
['order_id', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value']

=== order_payments ===
['order_id', 'payment_sequential', 'payment_type', 'payment_installments', 'payment_value']

=== order_reviews ===
['review_id', 'order_id', 'review_score', 'review_comment_title', 'review_comment_message', 'review_creation_date', 'review_answer_timestamp']

=== orders ===
['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_d

In [70]:
# Load all tables into DataFrames
df_product_category = pd.read_sql_query("SELECT * FROM product_category_name_translation", conn)
df_sellers = pd.read_sql_query("SELECT * FROM sellers", conn)
df_customers = pd.read_sql_query("SELECT * FROM customers", conn)
df_geolocation = pd.read_sql_query("SELECT * FROM geolocation", conn)
df_order_items = pd.read_sql_query("SELECT * FROM order_items", conn)
df_order_payments = pd.read_sql_query("SELECT * FROM order_payments", conn)
df_order_reviews = pd.read_sql_query("SELECT * FROM order_reviews", conn)
df_orders = pd.read_sql_query("SELECT * FROM orders", conn)
df_products = pd.read_sql_query("SELECT * FROM products", conn)
df_leads_qualified = pd.read_sql_query("SELECT * FROM leads_qualified", conn)
df_leads_closed = pd.read_sql_query("SELECT * FROM leads_closed", conn)


# Print the shape of each DataFrame to verify loading
print("Shapes of loaded DataFrames:")
print(f"Product Category: {df_product_category.shape}")
print(f"Sellers: {df_sellers.shape}")
print(f"Customers: {df_customers.shape}")
print(f"Geolocation: {df_geolocation.shape}")
print(f"Order Items: {df_order_items.shape}")
print(f"Order Payments: {df_order_payments.shape}")
print(f"Order Reviews: {df_order_reviews.shape}")
print(f"Orders: {df_orders.shape}")
print(f"Products: {df_products.shape}")
print(f"Leads Qualified: {df_leads_qualified.shape}")
print(f"Leads Closed: {df_leads_closed.shape}")

Shapes of loaded DataFrames:
Product Category: (71, 2)
Sellers: (3095, 4)
Customers: (99441, 5)
Geolocation: (1000163, 5)
Order Items: (112650, 7)
Order Payments: (103886, 5)
Order Reviews: (99224, 7)
Orders: (99441, 8)
Products: (32951, 9)
Leads Qualified: (8000, 4)
Leads Closed: (842, 14)


## Data Preprocessing pipeline

In [6]:
# create a function for preprocessing

def analyze_table(table_name):
    """
    Comprehensive analysis of a single table
    """
    print(f"\n{'='*50}")
    print(f"Analyzing table: {table_name}")
    print(f"{'='*50}")
    
    # Read table
    df = pd.read_sql_query(f"SELECT * FROM {table_name}", conn)
    
    # 1. Basic Information
    print("\n1. Basic Information:")
    print(f"Shape: {df.shape}")
    print("\nData Types:")
    print(df.dtypes)
    
    # 2. Missing Values
    print("\n2. Missing Values Analysis:")
    missing = df.isnull().sum()
    missing_percent = (df.isnull().sum() / len(df)) * 100
    missing_data = pd.concat([missing, missing_percent], axis=1)
    missing_data.columns = ['Missing Count', 'Missing Percent']
    print(missing_data[missing_data['Missing Count'] > 0])
    
    # 3. Duplicate Rows
    duplicates = df.duplicated().sum()
    print(f"\n3. Duplicate Rows: {duplicates}")
    
    # 4. Basic Statistics for Numeric Columns
    if not df.select_dtypes(include=[np.number]).empty:
        print("\n4. Basic Statistics for Numeric Columns:")
        print(df.describe())
    
    # 5. Unique Values in Each Column
    print("\n5. Unique Values Count and Sample Values:")
    for column in df.columns:
        unique_count = df[column].nunique()
        sample_values = df[column].dropna().sample(min(3, unique_count)).tolist()
        print(f"\n{column}:")
        print(f"- Unique values: {unique_count}")
        print(f"- Sample values: {sample_values}")
        
    return df
   

## Data exporation

## 1. Product category

In [7]:
c

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

In [8]:
# Explore data
df_product_category = analyze_table('product_category_name_translation')


Analyzing table: product_category_name_translation

1. Basic Information:
Shape: (71, 2)

Data Types:
product_category_name            object
product_category_name_english    object
dtype: object

2. Missing Values Analysis:
Empty DataFrame
Columns: [Missing Count, Missing Percent]
Index: []

3. Duplicate Rows: 0

5. Unique Values Count and Sample Values:

product_category_name:
- Unique values: 71
- Sample values: ['bebes', 'utilidades_domesticas', 'fashion_roupa_infanto_juvenil']

product_category_name_english:
- Unique values: 71
- Sample values: ['diapers_and_hygiene', 'art', 'small_appliances_home_oven_and_coffee']


### Observation
* The data is clean, with no missing or duplicate records, making it ready for analysis.
The dataset appears to be structured for easy category mapping or translation reference in multilingual contexts, possibly aiding in product classification or cross-language analysis.

## 2. Sellers

In [9]:
# Take a peak at the  sellers dataframe
df_sellers.head()

seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

In [10]:
# Reload sellers from your original SQL query
df_sellers = pd.read_sql_query("SELECT * FROM sellers", conn)

In [11]:
# Explore data
df_sellers = analyze_table('sellers')


Analyzing table: sellers

1. Basic Information:
Shape: (3095, 4)

Data Types:
seller_id                 object
seller_zip_code_prefix     int64
seller_city               object
seller_state              object
dtype: object

2. Missing Values Analysis:
Empty DataFrame
Columns: [Missing Count, Missing Percent]
Index: []

3. Duplicate Rows: 0

4. Basic Statistics for Numeric Columns:
       seller_zip_code_prefix
count             3095.000000
mean             32291.059451
std              32713.453830
min               1001.000000
25%               7093.500000
50%              14940.000000
75%              64552.500000
max              99730.000000

5. Unique Values Count and Sample Values:

seller_id:
- Unique values: 3095
- Sample values: ['9f505651f4a6abe901a56cdc21508025', '20b54c376b794ed028df09a3cd88e8dc', 'ebe3dabc4955c9b0d695e7bda38bf5c3']

seller_zip_code_prefix:
- Unique values: 2246
- Sample values: [89160, 29600, 95013]

seller_city:
- Unique values: 611
- Sample values: ['i

### Observations

The data is clean but needs further analysis and standardization

Urban Clustering: The presence of multiple sellers from the same city hints at seller concentration in major metropolitan areas, relevant for logistical optimisations.
Potential for Geo-Spatial Analysis: The combination of seller_zip_code_prefix, seller_city, and seller_state can be leveraged for mapping seller density, shipping route analysis, and regional performance metrics.
Zip Code Variability: The wide range of zip codes might impact delivery times and shipping costs, especially in operational efficiency analyses.


### Clean df_sellers 

In [12]:
# Look at seller_city distribution
print("=== City Analysis ===")
print(df_sellers['seller_city'].value_counts().head(10))
print("\nTotal unique cities:", df_sellers['seller_city'].nunique())

# Look at seller_state distribution
print("\n=== State Analysis ===")
print(df_sellers['seller_state'].value_counts())
print("\nTotal unique states:", df_sellers['seller_state'].nunique())

# Look at zip_code patterns
print("\n=== Zip Code Analysis ===")
print("Zip code length distribution:")
print(df_sellers['seller_zip_code_prefix'].astype(str).str.len().value_counts())
print("\nSample of different zip codes:")
print(df_sellers['seller_zip_code_prefix'].sample(10))

=== City Analysis ===
seller_city
sao paulo         694
curitiba          127
rio de janeiro     96
belo horizonte     68
ribeirao preto     52
guarulhos          50
ibitinga           49
santo andre        45
campinas           41
maringa            40
Name: count, dtype: int64

Total unique cities: 611

=== State Analysis ===
seller_state
SP    1849
PR     349
MG     244
SC     190
RJ     171
RS     129
GO      40
DF      30
ES      23
BA      19
CE      13
PE       9
PB       6
RN       5
MS       5
MT       4
RO       2
SE       2
PI       1
AC       1
MA       1
AM       1
PA       1
Name: count, dtype: int64

Total unique states: 23

=== Zip Code Analysis ===
Zip code length distribution:
seller_zip_code_prefix
5    2068
4    1027
Name: count, dtype: int64

Sample of different zip codes:
792     13800
2323    88303
2434    14940
2225    13908
1985     8050
1862    13690
548     83075
1179    14940
2852     1201
2815     5849
Name: seller_zip_code_prefix, dtype: int64


In [13]:
# 1. Standardize city names
df_sellers['seller_city'] = df_sellers['seller_city'].str.title()

# 2. Standardize zip codes to 5 digits
df_sellers['seller_zip_code_prefix'] = df_sellers['seller_zip_code_prefix'].astype(str).str.zfill(5)


In [14]:
# view df_sellers after cleaning 
df_sellers.head()

seller_id seller_zip_code_prefix        seller_city  \
0  3442f8959a84dea7ee197c632cb2df15                  13023           Campinas   
1  d1b65fc7debc3361ea86b5f14c68d2e2                  13844         Mogi Guacu   
2  ce3ad9de960102d0677a81f5d0bb7b2d                  20031     Rio De Janeiro   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                  04195          Sao Paulo   
4  51a04a8a6bdcb23deccc82b0b80742cf                  12914  Braganca Paulista   

  seller_state  
0           SP  
1           SP  
2           RJ  
3           SP  
4           SP

### 3. Customers

In [15]:
# Take a peak at the customers dataframe
df_customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [16]:
# Explore data
df_customers = analyze_table('customers')


Analyzing table: customers

1. Basic Information:
Shape: (99441, 5)

Data Types:
customer_id                 object
customer_unique_id          object
customer_zip_code_prefix     int64
customer_city               object
customer_state              object
dtype: object

2. Missing Values Analysis:
Empty DataFrame
Columns: [Missing Count, Missing Percent]
Index: []

3. Duplicate Rows: 0

4. Basic Statistics for Numeric Columns:
       customer_zip_code_prefix
count              99441.000000
mean               35137.474583
std                29797.938996
min                 1003.000000
25%                11347.000000
50%                24416.000000
75%                58900.000000
max                99990.000000

5. Unique Values Count and Sample Values:

customer_id:
- Unique values: 99441
- Sample values: ['b10420d1fe67d729669d0f2562333884', 'a08d203f34f5d5f55ad0dd27c6481588', 'ebb7a2942172efa549f149c74e43c3d6']

customer_unique_id:
- Unique values: 96096
- Sample values: ['d507e084dc5

### Observation

Repeat Customers: The difference between customer_id and customer_unique_id counts suggests customer retention and repeat purchase behavior, which is crucial for churn prediction and loyalty analysis.
Regional Coverage: Broad geographic spread allows for effective segmentation by state or city, supporting marketing strategies and logistics optimization.
Potential for Demographic Analysis: Although limited to location data, combining with external demographic datasets could enhance customer segmentation efforts.


## Clean df_customers

In [17]:
# Look at customer_unique_id duplicates
print("\nCustomers with multiple entries:")
print(df_customers['customer_unique_id'].value_counts().head(10))


Customers with multiple entries:
customer_unique_id
8d50f5eadf50201ccdcedfb9e2ac8455    17
3e43e6105506432c953e165fb2acf44c     9
1b6c7548a2a1f9037c1fd3ddfed95f33     7
ca77025e7201e3b30c44b472ff346268     7
6469f99c1f9dfae7733b25662e7f1782     7
63cfc61cee11cbe306bff5857d00bfe4     6
47c1a3033b8b77b3ab6e109eb4d5fdf3     6
12f5d6e1cbf93dafd9dcc19095df0b3d     6
de34b16117594161a6a89c50b289d35a     6
dc813062e0fc23409cd255f7f53c7074     6
Name: count, dtype: int64


In [18]:
# Look at full details of customer with 17 entries
print(df_customers[df_customers['customer_unique_id'] == '8d50f5eadf50201ccdcedfb9e2ac8455'])


                            customer_id                customer_unique_id  \
14186  1bd3585471932167ab72a84955ebefea  8d50f5eadf50201ccdcedfb9e2ac8455   
15321  a8fabc805e9a10a3c93ae5bff642b86b  8d50f5eadf50201ccdcedfb9e2ac8455   
16654  897b7f72042714efaa64ac306ba0cafc  8d50f5eadf50201ccdcedfb9e2ac8455   
36122  b2b13de0770e06de50080fea77c459e6  8d50f5eadf50201ccdcedfb9e2ac8455   
38073  42dbc1ad9d560637c9c4c1533746f86d  8d50f5eadf50201ccdcedfb9e2ac8455   
40141  dfb941d6f7b02f57a44c3b7c3fefb44b  8d50f5eadf50201ccdcedfb9e2ac8455   
48614  65f9db9dd07a4e79b625effa4c868fcb  8d50f5eadf50201ccdcedfb9e2ac8455   
52574  1c62b48fb34ee043310dcb233caabd2e  8d50f5eadf50201ccdcedfb9e2ac8455   
58707  a682769c4bc10fc6ef2101337a6c83c9  8d50f5eadf50201ccdcedfb9e2ac8455   
67996  6289b75219d757a56c0cce8d9e427900  8d50f5eadf50201ccdcedfb9e2ac8455   
72745  3414a9c813e3ca02504b8be8b2deb27f  8d50f5eadf50201ccdcedfb9e2ac8455   
74510  0e4fdc084a6b9329ed55d62dcd653ccf  8d50f5eadf50201ccdcedfb9e2ac8455   

## Observation
The customers table (99,441 rows, 5 columns) reveals a well-structured e-commerce customer tracking system:

Customer ID System:


Each transaction gets a unique customer_id
customer_unique_id tracks repeat customers
Example: One customer (8d50f5...) has 17 different transaction IDs but same location


Geographic Distribution:


27 states represented (all Brazilian states + Federal District)
Expected concentration in populous states (SP: 41,746, RJ: 12,852)
Lower numbers in less populated states (RR: 46, AP: 68)


Data Quality:


No missing values
No duplicate records
Consistent format across entries
No cleaning needed except standardization of city names and zip codes

In [19]:
# Standardize city names to Title Case
df_customers['customer_city'] = df_customers['customer_city'].str.title()

In [20]:
#  Standardize zip codes to 5 digits
df_customers['customer_zip_code_prefix'] = df_customers['customer_zip_code_prefix'].astype(str).str.zfill(5)


In [21]:
# take a look at the cleaned data
df_customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

  customer_zip_code_prefix          customer_city customer_state  
0                    14409                 Franca             SP  
1                    09790  Sao Bernardo Do Campo             SP  
2                    01151              Sao Paulo             SP  
3                    08775        Mogi Das Cruzes             SP  
4                    13056               Campinas             SP

## 4. Geolocation

In [22]:
# Take a peak at the geolocation dataframe
df_geolocation.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

In [23]:
# Explore data
df_geolocation = analyze_table('geolocation')


Analyzing table: geolocation

1. Basic Information:
Shape: (1000163, 5)

Data Types:
geolocation_zip_code_prefix      int64
geolocation_lat                float64
geolocation_lng                float64
geolocation_city                object
geolocation_state               object
dtype: object

2. Missing Values Analysis:
Empty DataFrame
Columns: [Missing Count, Missing Percent]
Index: []

3. Duplicate Rows: 261831

4. Basic Statistics for Numeric Columns:
       geolocation_zip_code_prefix  geolocation_lat  geolocation_lng
count                 1.000163e+06     1.000163e+06     1.000163e+06
mean                  3.657417e+04    -2.117615e+01    -4.639054e+01
std                   3.054934e+04     5.715866e+00     4.269748e+00
min                   1.001000e+03    -3.660537e+01    -1.014668e+02
25%                   1.107500e+04    -2.360355e+01    -4.857317e+01
50%                   2.653000e+04    -2.291938e+01    -4.663788e+01
75%                   6.350400e+04    -1.997962e+01    -

### Observation

Dense Coverage: The dataset has rich geographical diversity, supporting analyses like customer segmentation, logistics optimization, and delivery time estimation.
Potential for Geo-Mapping: Latitude and longitude values enable advanced spatial analyses, such as heatmaps or distance calculations.
Data Cleaning Needed: High duplicate count and outliers may require preprocessing for cleaner insights. however we have to check the duplicate rows pattern to understand why there are so many duplicate rows

In [24]:
# Check duplicate rows pattern
print("\nSample of duplicate records:")
df_geolocation.value_counts().head()


Sample of duplicate records:


geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  geolocation_city  geolocation_state
88220                        -27.102099       -48.629613       itapema           SC                   314
6414                         -23.495901       -46.874687       barueri           SP                   189
                             -23.490618       -46.869004       barueri           SP                   127
5145                         -23.506049       -46.717377       sao paulo         SP                   126
22620                        -23.005514       -43.375964       rio de janeiro    RJ                   102
Name: count, dtype: int64

## Observation
The "duplicates" are not errors:


They represent different coordinate points within same zip code areas
This is normal for geographic data where one zip code covers multiple locations
The variation in coordinates provides more precise location data

In [25]:
# Standardize city names to Title Case
df_geolocation['geolocation_city'] = df_geolocation['geolocation_city'].str.title()


In [26]:
# Standardize zip codes to 5 digits
df_geolocation['geolocation_zip_code_prefix'] = df_geolocation['geolocation_zip_code_prefix'].astype(str).str.zfill(5)

In [27]:
df_geolocation.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                       01037       -23.545621       -46.639292   
1                       01046       -23.546081       -46.644820   
2                       01046       -23.546129       -46.642951   
3                       01041       -23.544392       -46.639499   
4                       01035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        Sao Paulo                SP  
1        Sao Paulo                SP  
2        Sao Paulo                SP  
3        Sao Paulo                SP  
4        Sao Paulo                SP

## 5. order_items

In [28]:
# Take a peak at the order items dataframe
df_order_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [29]:
# Explore data
df_order_items = analyze_table('order_items')


Analyzing table: order_items

1. Basic Information:
Shape: (112650, 7)

Data Types:
order_id                object
order_item_id            int64
product_id              object
seller_id               object
shipping_limit_date     object
price                  float64
freight_value          float64
dtype: object

2. Missing Values Analysis:
Empty DataFrame
Columns: [Missing Count, Missing Percent]
Index: []

3. Duplicate Rows: 0

4. Basic Statistics for Numeric Columns:
       order_item_id          price  freight_value
count  112650.000000  112650.000000  112650.000000
mean        1.197834     120.653739      19.990320
std         0.705124     183.633928      15.806405
min         1.000000       0.850000       0.000000
25%         1.000000      39.900000      13.080000
50%         1.000000      74.990000      16.260000
75%         1.000000     134.900000      21.150000
max        21.000000    6735.000000     409.680000

5. Unique Values Count and Sample Values:

order_id:
- Unique v

### Observation
Potential Data Quality Issues:

Negative Bound for Price Outliers: The lower bound for price outliers is negative (-102.60), which is not logically possible for product prices. This may be due to incorrect outlier detection or data entry errors.
Key Insights:

Multiple Items per Order: Presence of order_item_id values greater than 1 suggests the need to aggregate data at the order_id level for certain analyses, like revenue per order.
Shipping Cost Variability: High variability in freight_value may impact profitability analyses and logistics optimization.
Price Distribution: Wide price range provides opportunities for segmentation (e.g., premium vs. budget products).
Action Required: Convert shipping_limit_date to datetime format for temporal analysis (e.g., delivery performance).

## Clean  order items

In [30]:
# Convert to datetime
df_order_items['shipping_limit_date'] = pd.to_datetime(df_order_items['shipping_limit_date'])

In [31]:
# Check very low prices
print(df_order_items[df_order_items['price'] < 1.0])

                               order_id  order_item_id  \
27652  3ee6513ae7ea23bdfab5b9ab60bffcb5              1   
48625  6e864b3f0ec71031117ad4cf46b7f2a1              1   
87081  c5bdd8ef3c0ec420232e668302179113              2   

                             product_id                         seller_id  \
27652  8a3254bee785a526d548a81a9bc3c9be  96804ea39d96eb908e7c3afdb671bb9e   
48625  8a3254bee785a526d548a81a9bc3c9be  96804ea39d96eb908e7c3afdb671bb9e   
87081  8a3254bee785a526d548a81a9bc3c9be  96804ea39d96eb908e7c3afdb671bb9e   

      shipping_limit_date  price  freight_value  
27652 2018-05-04 03:55:26   0.85          18.23  
48625 2018-05-02 20:30:34   0.85          18.23  
87081 2018-05-07 02:55:22   0.85          22.30  


In [32]:
# Check zero freight orders
print(df_order_items[df_order_items['freight_value'] == 0])

                                order_id  order_item_id  \
114     00404fa7a687c8c44ca69d42695aae73              1   
258     00a870c6c06346e85335524935c600c0              1   
483     011c899816ea29773525bd3322dbb6aa              1   
508     012b3f6ab7776a8ab3443a4ad7bef2e6              1   
509     012b3f6ab7776a8ab3443a4ad7bef2e6              2   
...                                  ...            ...   
111094  fc698f330ec7fb74859071cc6cb29772              1   
111497  fd4907109f6bac23f07064af84bec02d              1   
111649  fd95e4b85ebbb81853d4a6be3d61432b              1   
112182  fee19a0dc7358b6962a611cecf6a37b4              1   
112607  ffe73fc1d73e03fdb50e63903ddfe882              1   

                              product_id                         seller_id  \
114     53b36df67ebb7c41585e8d54d6772e08  7d13fca15225358621be4086e1eb0964   
258     aca2eb7d00ea1a7b8ebd4e68314663af  955fee9216a65b617aa5c0531780ce60   
483     53b36df67ebb7c41585e8d54d6772e08  7d13fca15225358

## observations 
1. Very consistent pattern
   
Different order IDs and dates
Same product-seller combination
Could be:

Sample/promotional item
Small accessory
Loss leader product

2. Notable Patterns:


Multiple sellers involved
Some recurring product IDs
Same shipping_limit_date for consecutive orders
Could indicate:

Free shipping promotions
Special campaigns
Local deliveries

This appears to be legitimate business practice rather than data error, as:

Small percentage of total orders
Normal price ranges
Structured shipping dates
Multiple sellers participating

In [33]:
# view cleaned order_items
df_order_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

  shipping_limit_date   price  freight_value  
0 2017-09-19 09:45:35   58.90          13.29  
1 2017-05-03 11:05:13  239.90          19.93  
2 2018-01-18 14:48:30  199.00          17.87  
3 2018-08-15 10:10:18   12.99          12.79  
4 2017-02-13 13:57:51  199.90          18.14

In [34]:
df_order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   order_id             112650 non-null  object        
 1   order_item_id        112650 non-null  int64         
 2   product_id           112650 non-null  object        
 3   seller_id            112650 non-null  object        
 4   shipping_limit_date  112650 non-null  datetime64[ns]
 5   price                112650 non-null  float64       
 6   freight_value        112650 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(1), object(3)
memory usage: 6.0+ MB


## 6. order_payments

In [35]:
# Take a peak at the order payments dataframe
df_order_payments.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [36]:
# Explore data
df_order_payments = analyze_table('order_payments')


Analyzing table: order_payments

1. Basic Information:
Shape: (103886, 5)

Data Types:
order_id                 object
payment_sequential        int64
payment_type             object
payment_installments      int64
payment_value           float64
dtype: object

2. Missing Values Analysis:
Empty DataFrame
Columns: [Missing Count, Missing Percent]
Index: []

3. Duplicate Rows: 0

4. Basic Statistics for Numeric Columns:
       payment_sequential  payment_installments  payment_value
count       103886.000000         103886.000000  103886.000000
mean             1.092679              2.853349     154.100380
std              0.706584              2.687051     217.494064
min              1.000000              0.000000       0.000000
25%              1.000000              1.000000      56.790000
50%              1.000000              1.000000     100.000000
75%              1.000000              4.000000     171.837500
max             29.000000             24.000000   13664.080000

5. Unique

### Observation

Dominant Single-Payment Pattern: Most transactions are one-time payments, which simplifies revenue recognition.
Payment Diversity: Multiple payment types and installment options provide opportunities for segmentation and customer preference analysis.
Outlier Relevance: High-value transactions could significantly influence revenue metrics and may require separate analysis to avoid skewing results.
Action Required: Convert payment_value to absolute values if any negatives are confirmed, and review zero-installment cases for validity.


In [37]:
print("Payment Types Distribution:")
print(df_order_payments['payment_type'].value_counts())

Payment Types Distribution:
payment_type
credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: count, dtype: int64


In [38]:
# Check zero payment values
print("Zero payment values:")
print(df_order_payments[df_order_payments['payment_value'] == 0])

# Check zero installments
print("\nZero installments:")
print(df_order_payments[df_order_payments['payment_installments'] == 0])

Zero payment values:
                                order_id  payment_sequential payment_type  \
19922   8bcbe01d44d147f901cd3192671144db                   4      voucher   
36822   fa65dad1b0e818e3ccc5cb0e39231352                  14      voucher   
43744   6ccb433e00daae1283ccc956189c82ae                   4      voucher   
51280   4637ca194b6387e2d538dc89b124b0ee                   1  not_defined   
57411   00b1cb0320190ca0daa2c88b35206009                   1  not_defined   
62674   45ed6e85398a87c253db47c2d9f48216                   3      voucher   
77885   fa65dad1b0e818e3ccc5cb0e39231352                  13      voucher   
94427   c8c528189310eaa44a745b8d9d26908b                   1  not_defined   
100766  b23878b3e8eb4d25a158f57d96331b18                   4      voucher   

        payment_installments  payment_value  
19922                      1            0.0  
36822                      1            0.0  
43744                      1            0.0  
51280                   

In [39]:
# Find orders with multiple payments
multiple_payments = df_order_payments['order_id'].value_counts()
print("Orders with multiple payments:")
print(multiple_payments[multiple_payments > 1].head())

# See details of a specific order with multiple payments
print("\nExample of order with multiple payments:")
print(df_order_payments[df_order_payments['order_id'] == multiple_payments[multiple_payments > 1].index[0]])

Orders with multiple payments:
order_id
fa65dad1b0e818e3ccc5cb0e39231352    29
ccf804e764ed5650cd8759557269dc13    26
285c2e15bebd4ac83635ccc563dc71f4    22
895ab968e7bb0d5659d16cd74cd1650c    21
fedcd9f7ccdc8cba3a18defedd1a5547    19
Name: count, dtype: int64

Example of order with multiple payments:
                                order_id  payment_sequential payment_type  \
4885    fa65dad1b0e818e3ccc5cb0e39231352                  27      voucher   
9985    fa65dad1b0e818e3ccc5cb0e39231352                   4      voucher   
14321   fa65dad1b0e818e3ccc5cb0e39231352                   1      voucher   
17274   fa65dad1b0e818e3ccc5cb0e39231352                   9      voucher   
19565   fa65dad1b0e818e3ccc5cb0e39231352                  10      voucher   
23074   fa65dad1b0e818e3ccc5cb0e39231352                   2      voucher   
24879   fa65dad1b0e818e3ccc5cb0e39231352                  25      voucher   
28330   fa65dad1b0e818e3ccc5cb0e39231352                   5      voucher   
2964

## observations and way forward

1. Conduct Thorough Exploratory Data Analysis (EDA)
Examine the Data Distribution:
I would start by analyzing the overall distribution of payment values and installments across all payment types. This includes:
Checking the frequency of zero values for different payment types (e.g., voucher vs. credit card).
Verifying whether zero installments for credit cards are isolated incidents or a broader pattern.
Compare with Similar Records:
I’d analyze similar transactions (e.g., other credit card payments) to see what installment values they have. For example, if most credit card transactions have an installment value of 1, then a zero could indicate a data entry error or a special case.
2. Validate with External Benchmarks and Domain Knowledge
Leverage Industry Standards:
In many systems, a single installment is typically recorded as “1.” The occurrence of “0” in credit card transactions might be an anomaly or a placeholder for “not applicable.” I’d compare this with standard practices in similar payment systems.

Assumption Check for Vouchers:
For voucher payments consistently showing a payment value of zero, I would check if this aligns with the understanding that vouchers might represent non-monetary transactions (e.g., discounts, credits, or redemption events).

3. Hypothesis Testing and Sensitivity Analysis
Create Hypotheses:
For instance:
Hypothesis A: Zero installments for credit card transactions are errors, and the intended value is 1.
Hypothesis B: Zero payment values for vouchers are by design, representing a different kind of transaction.
Test Impact on Key Metrics:
I would re-run the analysis under different assumptions (e.g., treating a zero installment as 1) and observe how sensitive the final metrics (like proxy CAC or conversion rates) are to these changes.
4. Implement Data Cleaning/Standardization Steps
Apply Corrections Where Justified:
If the analysis confirms that credit card payments should logically have at least one installment, I would recode zero values to 1. This should be done only if the data supports this correction and the impact is documented.

Flag or Document Anomalies:
For voucher transactions or any other discrepancies, I’d flag these records and include commentary in the analysis. This ensures that downstream users are aware of the decisions made.

5. Document Assumptions and Decisions Transparently
Create a Data Dictionary or Annotation:
Document the rationale behind each transformation and assumption. This documentation should include:
The observation that voucher payments have a zero value and that credit card payments sometimes have zero installments.
The decision to treat credit card zero installments as 1 (if the data supports it) or leave them as-is with appropriate notes.
Outline Limitations:
Clearly state that in the absence of stakeholder input or additional documentation, these decisions are based on industry norms and exploratory analysis. Future analyses should re-evaluate these assumptions if new information becomes available.
6. Proceed with the Analysis
Run the Adjusted Analysis:
With the cleaned and standardized data, continue with your primary objective (e.g., identifying the most cost-effective channels). Use conversion metrics, proxy CAC, or other relevant KPIs that can be reliably computed with the available data.

## 7. order_reviews

In [40]:
# Take a peak at the order reviews dataframe
df_order_reviews.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                 None   
1             5                 None   
2             5                 None   
3             5                 None   
4             5                 None   

                              review_comment_message review_creation_date  \
0                                               None  2018-01-18 00:00:00   
1                                               None  2018-03-10 00:00:00   
2                                               None  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

In [41]:
# Explore data
df_order_reviews = analyze_table('order_reviews')


Analyzing table: order_reviews

1. Basic Information:
Shape: (99224, 7)

Data Types:
review_id                  object
order_id                   object
review_score                int64
review_comment_title       object
review_comment_message     object
review_creation_date       object
review_answer_timestamp    object
dtype: object

2. Missing Values Analysis:
                        Missing Count  Missing Percent
review_comment_title            87656        88.341530
review_comment_message          58247        58.702532

3. Duplicate Rows: 0

4. Basic Statistics for Numeric Columns:
       review_score
count  99224.000000
mean       4.086421
std        1.347579
min        1.000000
25%        4.000000
50%        5.000000
75%        5.000000
max        5.000000

5. Unique Values Count and Sample Values:

review_id:
- Unique values: 98410
- Sample values: ['3b2081544c99c761c4a681bdf6425333', '05bbf469e420e24953561adec3f545d1', 'ba6c65519e0369dc9f73dcea58c93e11']

order_id:
- Unique 

### Observation

Positive Bias: The majority of reviews are positive, as reflected by the high mean and median scores.

Sentiment Analysis Potential: Despite missing comments, the available text data provides valuable insights into customer satisfaction and pain points.

Action Required: Convert date fields to datetime for time-based analyses and refine outlier detection for categorical ratings.

## Cleaning df_order_reviews

In [75]:
# Drop 'review_comment_title' column
df_order_reviews.drop(columns=['review_comment_title'], inplace=True)

# Drop rows where 'review_comment_message' is missing
df_order_reviews.dropna(subset=['review_comment_message'], inplace=True)

# Reset index after dropping rows
df_order_reviews.reset_index(drop=True, inplace=True)



In [43]:
# Standardize text fields by converting to lowercase and stripping extra whitespace
df_order_reviews['review_comment_title'] = df_order_reviews['review_comment_title'].str.lower().str.strip()
df_order_reviews['review_comment_message'] = df_order_reviews['review_comment_message'].str.lower().str.strip()

In [44]:
# Convert date columns to datetime
df_order_reviews['review_creation_date'] = pd.to_datetime(df_order_reviews['review_creation_date'])
df_order_reviews['review_answer_timestamp'] = pd.to_datetime(df_order_reviews['review_answer_timestamp'])

In [45]:
# view cleaned df_order_reviews
df_order_reviews.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4             no title   
1             5             no title   
2             5             no title   
3             5             no title   
4             5             no title   

                              review_comment_message review_creation_date  \
0                                         no message           2018-01-18   
1                                         no message           2018-03-10   
2                                         no message           2018-02-17   
3              recebi bem antes do prazo estipulado.           2017-04-21   
4  parabéns lojas lannister adorei comprar pela i...           2018-03-01   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

In [76]:
# view  df_order_reviews data types
df_order_reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40977 entries, 0 to 40976
Data columns (total 6 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                40977 non-null  object
 1   order_id                 40977 non-null  object
 2   review_score             40977 non-null  int64 
 3   review_comment_message   40977 non-null  object
 4   review_creation_date     40977 non-null  object
 5   review_answer_timestamp  40977 non-null  object
dtypes: int64(1), object(5)
memory usage: 1.9+ MB


## 8. orders

In [47]:
# Take a peak at the orders dataframe
df_orders .head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [48]:
# Explore data
df_orders = analyze_table('orders')


Analyzing table: orders

1. Basic Information:
Shape: (99441, 8)

Data Types:
order_id                         object
customer_id                      object
order_status                     object
order_purchase_timestamp         object
order_approved_at                object
order_delivered_carrier_date     object
order_delivered_customer_date    object
order_estimated_delivery_date    object
dtype: object

2. Missing Values Analysis:
                               Missing Count  Missing Percent
order_approved_at                        160         0.160899
order_delivered_carrier_date            1783         1.793023
order_delivered_customer_date           2965         2.981668

3. Duplicate Rows: 0

5. Unique Values Count and Sample Values:

order_id:
- Unique values: 99441
- Sample values: ['e992307f910844d84f796767fa1c557c', 'f91c5f1e1175df3a7bb0aea646ca91fd', '09badf53476a6bda72b4728e858a3433']

customer_id:
- Unique values: 99441
- Sample values: ['aa48045e1b356dad31ce9820e0170

### Observation
Missing Dates: Missing values in order_delivered_customer_date and order_delivered_carrier_date could signal incomplete transactions, canceled orders, or data entry gaps.
Date Consistency: Some records might have inconsistencies (e.g., orders marked as "delivered" without a corresponding order_delivered_customer_date), requiring validation.

#### Key Insights:

Delivery Performance Analysis: The combination of actual delivery dates vs. estimated delivery dates allows for evaluating delivery performance, delays, and customer satisfaction.

Order Lifecycle Tracking: The availability of multiple timestamps enables the analysis of order processing times—from purchase to approval, shipping, and final delivery.

Cancellation Insights: Missing delivery dates or approval timestamps can help identify patterns in canceled or problematic orders.
Action Items:

Convert Date Columns: Change all date-related columns to datetime format for accurate temporal analysis.

Investigate Missing Values: Especially in delivery-related fields to distinguish between canceled, returned, or delayed orders.

Calculate Order Durations: Derive new features like:
Order Processing Time: order_approved_at - order_purchase_timestamp
Shipping Time: order_delivered_customer_date - order_delivered_carrier_date
Delivery Delay: order_delivered_customer_date - order_estimated_delivery_date


## clean Orders

In [49]:
# Step 1: Convert date columns to datetime with error coercion.
date_columns = ['order_purchase_timestamp', 
                'order_approved_at', 
                'order_delivered_carrier_date', 
                'order_delivered_customer_date', 
                'order_estimated_delivery_date']

for col in date_columns:
    df_orders[col] = pd.to_datetime(df_orders[col])

In [50]:
#Print missing values summary.
print("Missing Values in orders table:")
print(df_orders.isnull().sum())

Missing Values in orders table:
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64


In [51]:
df_orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0                    2017-10-18  
1                    2018-08-13  
2                    2018-09-04  
3                    2017-12-15  
4                    2018-02-26

In [52]:
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  object        
 1   customer_id                    99441 non-null  object        
 2   order_status                   99441 non-null  object        
 3   order_purchase_timestamp       99441 non-null  datetime64[ns]
 4   order_approved_at              99281 non-null  datetime64[ns]
 5   order_delivered_carrier_date   97658 non-null  datetime64[ns]
 6   order_delivered_customer_date  96476 non-null  datetime64[ns]
 7   order_estimated_delivery_date  99441 non-null  datetime64[ns]
dtypes: datetime64[ns](5), object(3)
memory usage: 6.1+ MB


# NOTE
I would Leave the missing values and filter out when necessary. This preserves the integrity of your data without introducing potentially misleading imputation.

i can handle missing data in my computations.
The missingness itself might be informative (for example, orders missing a delivery date might have a particular status).
If i later discover that these missing values interfere with specific calculations (like computing delivery time intervals), i can always filter out or drop those rows at that stage.

## 9. products

In [53]:
# Take a peak at the products dataframe
df_products.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [54]:
# Explore data
df_products = analyze_table('products')


Analyzing table: products

1. Basic Information:
Shape: (32951, 9)

Data Types:
product_id                     object
product_category_name          object
product_name_lenght           float64
product_description_lenght    float64
product_photos_qty            float64
product_weight_g              float64
product_length_cm             float64
product_height_cm             float64
product_width_cm              float64
dtype: object

2. Missing Values Analysis:
                            Missing Count  Missing Percent
product_category_name                 610         1.851234
product_name_lenght                   610         1.851234
product_description_lenght            610         1.851234
product_photos_qty                    610         1.851234
product_weight_g                        2         0.006070
product_length_cm                       2         0.006070
product_height_cm                       2         0.006070
product_width_cm                        2         0.006070

3.

### Observations

* Unrealistic Dimension Values:
Zero values for weights and dimensions indicate potential data entry errors.
Extremely large dimensions (e.g., width of 118 cm) could be outliers or data errors.
Missing Category Information: Products without product_category_name may affect category-based analyses.
Key Insights:

Product Variety: The dataset includes a wide variety of products across different categories, useful for recommendation systems and segmentation.

* Action Items:

Data Cleaning:
Impute or flag missing values in key fields.
Handle unrealistic zero or extreme values in weights and dimensions.

Feature Engineering:
Calculate product volume (length × width × height).
Derive density metrics for logistical optimization.


## clean Products 

In [55]:

# Standardize product_category_name
df_products['product_category_name'] = df_products['product_category_name'].str.lower().str.strip()

In [56]:
# Fill missing product_category_name with the most frequent category (mode)
most_frequent_category = df_products['product_category_name'].mode()[0]
df_products['product_category_name'] = df_products['product_category_name'].fillna(most_frequent_category)


### Impute missing values for numeric columns using the median

In [57]:

# List of columns with ~1.85% missing values
cols_to_impute = ['product_name_lenght', 'product_description_lenght', 'product_photos_qty']
for col in cols_to_impute:
    median_val = df_products[col].median()
    df_products[col] = df_products[col].fillna(median_val)

# List of measurement columns with very few missing values
measurement_cols = ['product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm']
for col in measurement_cols:
    median_val = df_products[col].median()
    df_products[col] = df_products[col].fillna(median_val)

# Convert product_photos_qty to integer since it's meant to be a count
df_products['product_photos_qty'] = df_products['product_photos_qty'].astype(int)

# Verify cleaning
print("Data types and missing value counts after cleaning:")
print(df_products.info())
print(df_products.isnull().sum())

Data types and missing value counts after cleaning:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32951 non-null  object 
 2   product_name_lenght         32951 non-null  float64
 3   product_description_lenght  32951 non-null  float64
 4   product_photos_qty          32951 non-null  int32  
 5   product_weight_g            32951 non-null  float64
 6   product_length_cm           32951 non-null  float64
 7   product_height_cm           32951 non-null  float64
 8   product_width_cm            32951 non-null  float64
dtypes: float64(6), int32(1), object(2)
memory usage: 2.1+ MB
None
product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty

# Note
Given that this project involves tasks like personalization and product recommendation—where correctly identifying product categories is important—it is best to preserve the fact that the product category is missing rather than assuming it belongs to the most common category. It allows me to treat missingness as its own signal and avoid potentially skewing the category distribution.

## 10. leads_qualified

In [58]:
# Take a peak at the leads qualified dataframe
df_leads_qualified.head()

mql_id first_contact_date  \
0  dac32acd4db4c29c230538b72f8dd87d         2018-02-01   
1  8c18d1de7f67e60dbd64e3c07d7e9d5d         2017-10-20   
2  b4bc852d233dfefc5131f593b538befa         2018-03-22   
3  6be030b81c75970747525b843c1ef4f8         2018-01-22   
4  5420aad7fec3549a85876ba1c529bd84         2018-02-21   

                    landing_page_id          origin  
0  88740e65d5d6b056e0cda098e1ea6313          social  
1  007f9098284a86ee80ddeb25d53e0af8     paid_search  
2  a7982125ff7aa3b2054c6e44f9d28522  organic_search  
3  d45d558f0daeecf3cccdffe3c59684aa           email  
4  b48ec5f3b04e9068441002a19df93c6c  organic_search

In [59]:
# Explore data
df_leads_qualified = analyze_table('leads_qualified')


Analyzing table: leads_qualified

1. Basic Information:
Shape: (8000, 4)

Data Types:
mql_id                object
first_contact_date    object
landing_page_id       object
origin                object
dtype: object

2. Missing Values Analysis:
        Missing Count  Missing Percent
origin             60             0.75

3. Duplicate Rows: 0

5. Unique Values Count and Sample Values:

mql_id:
- Unique values: 8000
- Sample values: ['09209177cb567dae63ade4fea6c65957', 'c4baf6469a476f27028b343c4aaa0883', '68453fe4a8fe4478c75986c5da65af22']

first_contact_date:
- Unique values: 336
- Sample values: ['2018-05-22', '2018-05-29', '2018-01-16']

landing_page_id:
- Unique values: 495
- Sample values: ['22c29808c4f815213303f8933030604c', 'b48ec5f3b04e9068441002a19df93c6c', '21a104c3279681d9856b8a48723c4fee']

origin:
- Unique values: 10
- Sample values: ['organic_search', 'organic_search', 'organic_search']


### Observations
Unique Identifiers and Diversity:

mql_id: 8,000 unique values, indicating every lead is distinct.
first_contact_date: 336 unique dates, showing leads were acquired over an extended period, which allows for temporal trend analysis.
landing_page_id: 495 unique landing pages, suggesting a diverse range of campaigns or landing page variations.
origin: 10 unique lead sources, including:
Sample Origins: 'paid_search', 'unknown', 'organic_search'.
Potential Data Quality Issues:

Missing origin Values: While minimal, these missing values may affect attribution modeling or CAC analysis if not addressed.

unknown as an Origin: This category might represent either genuinely untracked leads or data entry inconsistencies. It could be grouped with missing values for analysis.

Key Insights:

Lead Source Diversity: A variety of acquisition channels (e.g., paid search, organic search) allows for detailed attribution analysis to determine the most effective marketing strategies.
Landing Page Effectiveness: The large number of unique landing_page_id entries provides an opportunity to assess which landing pages convert leads most effectively.

Temporal Trends: The spread of first_contact_date values enables the analysis of lead acquisition trends over time.

Action Items:

Convert first_contact_date to datetime for better handling of time-series analysis.
Impute or handle missing origin values:
Consider labeling missing values as 'unknown' if appropriate, or treating them separately in analyses.
Analyze Lead Source Impact: Compare conversion rates across different origin categories to identify high-performing acquisition channels.

## Clean leads_ qualified

In [60]:
# Convert first_contact_date to datetime
df_leads_qualified['first_contact_date'] = pd.to_datetime(df_leads_qualified['first_contact_date'])


In [61]:
# Fill missing values with the most frequent origin (mode)
most_frequent_origin = df_leads_qualified['origin'].mode()[0]
df_leads_qualified['origin'] = df_leads_qualified['origin'].fillna(most_frequent_origin)


In [62]:
# Check the cleaned data
print("Data types and missing values after cleaning:")
print(df_leads_qualified.info())
print("\nUnique values in 'origin':")
print(df_leads_qualified['origin'].unique())

Data types and missing values after cleaning:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   mql_id              8000 non-null   object        
 1   first_contact_date  8000 non-null   datetime64[ns]
 2   landing_page_id     8000 non-null   object        
 3   origin              8000 non-null   object        
dtypes: datetime64[ns](1), object(3)
memory usage: 250.1+ KB
None

Unique values in 'origin':
['social' 'paid_search' 'organic_search' 'email' 'unknown' 'referral'
 'direct_traffic' 'display' 'other_publicities' 'other']


## 11. leads_closed 

In [63]:
# Take a peak at the leads closed dataframe
df_leads_closed .head()

mql_id                         seller_id  \
0  5420aad7fec3549a85876ba1c529bd84  2c43fb513632d29b3b58df74816f1b06   
1  a555fb36b9368110ede0f043dfc3b9a0  bbb7d7893a450660432ea6652310ebb7   
2  327174d3648a2d047e8940d7d15204ca  612170e34b97004b3ba37eae81836b4c   
3  f5fee8f7da74f4887f5bcae2bafb6dd6  21e1781e36faf92725dde4730a88ca0f   
4  ffe640179b554e295c167a2f6be528e0  ed8cb7b190ceb6067227478e48cf8dde   

                             sdr_id                             sr_id  \
0  a8387c01a09e99ce014107505b92388c  4ef15afb4b2723d8f3d81e51ec7afefe   
1  09285259593c61296eef10c734121d5b  d3d1e91a157ea7f90548eef82f1955e3   
2  b90f87164b5f8c2cfa5c8572834dbe3f  6565aa9ce3178a5caf6171827af3a9ba   
3  56bf83c4bb35763a51c2baab501b4c67  d3d1e91a157ea7f90548eef82f1955e3   
4  4b339f9567d060bcea4f5136b9f5949e  d3d1e91a157ea7f90548eef82f1955e3   

              won_date business_segment      lead_type lead_behaviour_profile  \
0  2018-02-26 19:58:54              pet  online_medium                    cat   
1  2018-05-08 20:17:59  car_accessories       industry                  eagle   
2  2018-06-05 17:27:23  home_appliances     online_big                    cat   
3  2018-01-17 13:51:03       food_drink   online_small                   None   
4  2018-07-03 20:17:45  home_appliances       industry                   wolf   

   has_company  has_gtin average_stock business_type  \
0          NaN       NaN          None      reseller   
1          NaN       NaN          None      reseller   
2          NaN       NaN          None      reseller   
3          NaN       NaN          None      reseller   
4          NaN       NaN          None  manufacturer   

   declared_product_catalog_size  declared_monthly_revenue  
0                            NaN                       0.0  
1                            NaN                       0.0  
2                            NaN                       0.0  
3                            NaN                       0.0  
4                            NaN                       0.0

In [64]:
# explore the data
df_leads_closed = analyze_table('leads_closed')


Analyzing table: leads_closed

1. Basic Information:
Shape: (842, 14)

Data Types:
mql_id                            object
seller_id                         object
sdr_id                            object
sr_id                             object
won_date                          object
business_segment                  object
lead_type                         object
lead_behaviour_profile            object
has_company                      float64
has_gtin                         float64
average_stock                     object
business_type                     object
declared_product_catalog_size    float64
declared_monthly_revenue         float64
dtype: object

2. Missing Values Analysis:
                               Missing Count  Missing Percent
business_segment                           1         0.118765
lead_type                                  6         0.712589
lead_behaviour_profile                   177        21.021378
has_company                              779       

### observation
Potential Data Quality Issues:

High Missingness: The high rate of missing data in key columns (like company information and product catalog size) limits analysis depth.

Inconsistent Data Types: average_stock is an object with ranges (e.g., '5-20'), which may need conversion to numerical ranges for quantitative analysis.
Outlier Detection Errors: Some outlier bounds are illogical, such as negative lower bounds for catalog size, which should always be positive.
Key Insights:

Company Association Matters: Leads with known companies and GTINs seem more likely to convert, though data is sparse.
Zero Revenue Concern: Many leads declare zero monthly revenue, possibly due to data entry gaps or early-stage businesses.
Sales Rep Performance: The limited number of SDRs and SRs enables focused performance analysis.
Action Items:

Handle Missing Values: Consider imputation strategies or flag missing data for further investigation.

Data Type Conversion:
Convert won_date to datetime.

Parse average_stock ranges into numerical format for better analysis.
Outlier Handling: Review and correct outlier detection methods, especially where negative bounds are illogical.


#### The Outlier detection code was removed, each dataframe will be treated based on the type of analysis i want to perform

## clean leads_closed

In [65]:
#  Convert won_date to datetime
df_leads_closed['won_date'] = pd.to_datetime(df_leads_closed['won_date'], errors='raise')

In [66]:
# Fill missing values for low-missingness categorical columns with mode
for col in ['business_segment', 'lead_type', 'business_type']:
    df_leads_closed[col] = df_leads_closed[col].fillna(df_leads_closed[col].mode()[0])

# Fill moderate-high missingness column with 'Unclassified'
df_leads_closed['lead_behaviour_profile'] = df_leads_closed['lead_behaviour_profile'].fillna('Unclassified')


# Handle high missingness columns
# Drop columns that are over 90% missing (has_company, has_gtin, average_stock, declared_product_catalog_size)
df_leads_closed = df_leads_closed.drop(columns=['has_company', 'has_gtin', 'average_stock', 'declared_product_catalog_size'])


In [67]:
# Check the cleaned data
print("Cleaned leads_closed data types and missing values:")
print(df_leads_closed.info())
print("\nMissing values summary:")
print(df_leads_closed.isnull().sum())
print("\nFirst 5 rows:")
print(df_leads_closed.head())

Cleaned leads_closed data types and missing values:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 842 entries, 0 to 841
Data columns (total 10 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   mql_id                    842 non-null    object        
 1   seller_id                 842 non-null    object        
 2   sdr_id                    842 non-null    object        
 3   sr_id                     842 non-null    object        
 4   won_date                  842 non-null    datetime64[ns]
 5   business_segment          842 non-null    object        
 6   lead_type                 842 non-null    object        
 7   lead_behaviour_profile    842 non-null    object        
 8   business_type             842 non-null    object        
 9   declared_monthly_revenue  842 non-null    float64       
dtypes: datetime64[ns](1), float64(1), object(8)
memory usage: 65.9+ KB
None

Missing values summar

### Consideration for declared_monthly_revenue:
Since declared_monthly_revenue is consistently 0.0, it might not add any informative variance for modeling. Depending on my analysis, i could consider dropping it later if it doesn't contribute to my predictive or segmentation tasks.

# Save all data frames to csv files for further use 

In [68]:

# 1. Product Category Name Translation
df_product_category.to_csv("product_category_name_translation_cleaned.csv", index=False)

# 2. Sellers
df_sellers.to_csv("sellers_cleaned.csv", index=False)

# 3. Customers
df_customers.to_csv("customers_cleaned.csv", index=False)

# 4. Geolocation
df_geolocation.to_csv("geolocation_cleaned.csv", index=False)

# 5. Order Items
df_order_items.to_csv("order_items_cleaned.csv", index=False)

# 6. Order Payments
df_order_payments.to_csv("order_payments_cleaned.csv", index=False)

# 7. Order Reviews
df_order_reviews.to_csv("order_reviews_cleaned.csv", index=False)

# 8. Orders
df_orders.to_csv("orders_cleaned.csv", index=False)

# 9. Products
df_products.to_csv("products_cleaned.csv", index=False)

# 10. Leads Qualified
df_leads_qualified.to_csv("leads_qualified_cleaned.csv", index=False)

# 11. Leads Closed
df_leads_closed.to_csv("leads_closed_cleaned.csv", index=False)
